build(cuda): register compute capability cfgs and optional legacy BF16/FP8 emulation

haricot opened the pull request:
tested and works with:
```
nvidia-smi   --query-gpu="compute_cap"  --format=csv
compute_cap
6.1
```
```
cargo run -F "cuda,cudnn" --example llama  --release -- --model-id meta-llama/Llama-3.2-1B-Instruct --temperature 0.1 --which v32-1b-instruct --seed 42 --dtype bf16 --prompt
```

related https://github.com/EricLBuehler/candle/pull/57

haricot commented:
You are certainly right because I do not encounter the condition __CUDA_ARCH__ >= 800 in this control flow in my case but this must possibly cause a function cuda error already present, I will review this.

LaurentMazare commented:
I think the current code is likely to result in lots of compile failures with cuda compute cap >= 8.0.
Would it work to just replace the checks for compute cap 800 with 530? If you can make the changes on an old cuda toolkit, I can also test them on hardware that is past 800.

haricot commented:
thin with the last commit I get:
cargo run -F "cuda,cudnn" --example llama  --release -- --model-id meta-llama/Llama-3.2-1B-Instruct --temperature 0.1 --which v32-1b-instruct --seed 42 --dtype bf16 --prompt "'tell me something joke?'"
Error: A weight is invalid in distribution

ok f16, f32

haricot commented:
I hope that the latest additions will allow to work on ` __CUDA_ARCH__ >=800`.
Then I think it logical to propose more tests by considering if possible a full support of the BF16 type for `__CUDA_ARCH__ >=530` .

haricot commented on candle-kernels/src/reduce.cu:
```diff
@@ -571,15 +571,57 @@ fast_argmax(const size_t src_numel, const size_t el_to_sum_per_block,
+#if __CUDA_ARCH__ >= 800
+SUM_OP(__nv_bfloat16, sum_bf16)
+#elif __CUDA_ARCH__ >= 530 &&  __CUDA_ARCH__ < 800
+#define SUM_BF16_OP(FN_NAME)                                                   \
```
without custom `SUM_BF16_OP`  for  `__CUDA_ARCH__ >= 530 && __CUDA_ARCH__ < 800`  I get:
```
  --- stderr
  src/reduce.cu(623): error: no instance of overloaded function "atomicAdd" matches the argument list
              argument types are: (__nv_bfloat16 *, const __nv_bfloat16)
    extern "C" __attribute__((global)) void sum_bf16( const size_t numel, const size_t num_dims, const size_t num_sum_dims, const size_t *info, const __nv_bfloat16 *inp, __nv_bfloat16 *out) { const size_t *dims = info; const size_t *strides = info + num_dims; const size_t *sum_dims_l = info + 2 * num_dims; const size_t *sum_dims_s = info + 2 * num_dims + num_sum_dims; if (is_contiguous(num_dims, dims, strides)) { for (unsigned int i = blockIdx.x * blockDim.x + threadIdx.x; i < numel; i += blockDim.x * gridDim.x) { size_t dst_index = i; for (unsigned int nd = 0; nd < num_sum_dims; ++nd) { size_t stride = sum_dims_s[nd]; size_t pre = dst_index / stride; size_t post = dst_index % stride; dst_index = (pre / sum_dims_l[nd]) * stride + post; } atomicAdd(out + dst_index, inp[i]); } } else { for (unsigned int i = blockIdx.x * blockDim.x + threadIdx.x; i < numel; i += blockDim.x * gridDim.x) { unsigned strided_i = get_strided_index(i, num_dims, dims, strides); size_t dst_index = i; for (unsigned int nd = 0; nd < num_sum_dims; ++nd) { size_t stride = sum_dims_s[nd]; size_t pre = dst_index / stride; size_t post = dst_index % stride; dst_index = (pre / sum_dims_l[nd]) * stride + post; } atomicAdd(out + dst_index, inp[strided_i]); } } }
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ^
  src/compatibility.cuh(42): note #3326-D: function "atomicAdd(__half *, __half)" does not match because argument #1 does not match parameter
```

ivarflakstad commented on candle-kernels/src/affine.cu:
```diff
@@ -28,11 +28,11 @@ extern "C" __global__ void FN_NAME(  \
 } \
 
-#if __CUDA_ARCH__ >= 800
+#if __CUDA_ARCH__ >= 800 || (__CUDA_ARCH__ >= 530 && __CUDA_ARCH__ < 800)
```
I don't mind the having the atomic compare and swap impl for atomicAdd for cards that don't support it natively. `dummy_bf16.cuh` is for bfloat16, not float16 (half).

But that's not what I'm trying to address.
I'm simply saying that the cuda arch if-statements in the PR are incorrect.

haricot replied:
Necessary because if __CUDA_ARCH__ is 900, this is native
```
#if __CUDA_ARCH__ >= 800
 ...
#elif (__CUDA_ARCH__ >= 530 && __CUDA_ARCH__ < 800) 
.. 
#endif

```

haricot replied:
You are right to point out that this is not explicitly clear. I included that so as not to erase the native logic. If you agree, I would prefer this:
```
#if __CUDA_ARCH__ >= 800
 ...
#elif USE_BF16_FALLBACK  
.. 
#endif

```
or 
```
#if __CUDA_ARCH__ >= 800 || USE_BF16_FALLBACK
 // instead of __CUDA_ARCH__ >= 540  which loses information about the native limit.
```

_half  atomicAdd is necessary for:
src/reduce.cu — sum_f16 
src/indexing.cu — index_add / scatter_add kernels If multiple threads write to the same index, an atomic addition or block aggregation strategy is required.

You are righit cast f32 to bf16 or cast bf16 to f32 don't use _half atomicAdd.

I'm trying to understand the nuance with dummy_bf16.

haricot replied:
but I think there is a problem if we use several different GPU compute capabilities, a single binary compiled for one capability may not work or lose optimization on other GPU.
We might have to modify build.rs and add a dispatch. (compile two variants (native vs fallback) and do dispatch and launch the appropriate kernel variant (native vs fallback)).

ivarflakstad replied:
I expect a user to compile ptx on the machine they intend to use said ptx on

haricot commented:
I added `ALLOW_LEGACY_BF16` and `ALLOW_LEGACY_FP8` as well as [moe_hfma2](https://github.com/haricot/candle/blob/9808d4861d37f6956428ab0ca5a213e144cbf75b/candle-kernels/src/moe/moe_hfma2.cu#L249-L302) (wwma fallback solution) for CC < 700-800, and build like #3300

haricot committed:
fix(cuda): correct warp reductions in rmsnorm/layernorm kernels

    Use warp-local lane indices when interpreting __activemask() during
    warp XOR reductions, and guard the second-stage shared-memory reduction
    for partial warp counts.

haricot committed:
fix(cuda): robust reductions and unified BF16 math utilities

    - Fix regression in CUDA reduction kernels (max/min/sum) where results
      incorrectly returned 0.0 for all-negative inputs or small block sizes.
    - Use `__activemask()` and safe bounds checks in 'reduce.cu' to prevent
      inactive lanes from corrupting reductions with 0.0 or undefined values.
    - Modernize reduction identities using `::cuda::std::numeric_limits<T>`.
    - Refactor 'cuda_utils.cuh' to unify BF16 math functions, reducing code
      duplication between Ampere+ and Legacy paths.
    - Ensure consistent NaN propagation for BF16 `maxg`/`ming` on all GPU
      architectures by using specialized `__hmax_nan`/`__hmin_nan` helpers.
    - Add regression test 'max_reduction_stability' in 'tensor_tests.rs'

haricot committed:
fix(cuda): gate legacy BF16/FP8 kernels and optimize f16 unary ops

haricot committed:
fix(cuda): fall back from cuDNN conv1d launch failures

haricot committed:
feat(cuda): add MoE kernel dispatch and support contiguous offset views

haricot committed:
fix(cuda): fall back from cuDNN conv2d launch failures

haricot committed:
fix(cuda): zero-initialize MoE outputs and refine backend selection

haricot committed:
fix(cuda): tighten legacy atomicAdd fallbacks and BF16/FP8 regression tests

haricot committed:
chore(cuda): drop candle-kernels cudarc test harness

haricot committed:
test(nn): drop redundant f16 and bf16 sigmoid checks

haricot commented:
I force-pushed a cleaned up version of the branch and would appreciate another review pass.

The history is now split into focused commits for capability detection, legacy BF16/FP8 support, cuDNN fallbacks, MoE runtime changes, and regression coverage. I also removed the local `candle-kernels` MoE test harness and the redundant `sigmoid_f16` / `sigmoid_bf16` checks to keep the series easier to review.

The goal of this PR is to make Candle run more reliably on older NVIDIA GPUs by adding legacy BF16/FP8 fallback paths, improving capability detection, preserving numerical behavior where possible, and falling back from failing cuDNN convolution launches. This is mainly aimed at pre-Ampere cards.

Example invocation on older GPU using ALLOW_LEGACY all or bf16,fp8:
 `ALLOW_LEGACY="all" cargo run -F "cuda,cudnn" --example ...`

haricot committed:
fix logic legacy bf16/fp8 macro checks in kernels

sempervictus commented:
If you're going all the way with this @guoqingbao now has an emulated FP4 impl as well which works great on SM70 anyway :-)

haricot committed:
fix: add fallback bf16 min/max for pre-Ampere GPUs, allow legacy bf16 builds, and add debug symbols configuration

haricot committed:
build: update cudaforge dependency to specific git revision

haricot committed:
feat: add candle-tiled-attn package with CUDA support for tiled attention kernels

haricot committed:
tiled_attn attn test

haricot committed:
test: add CUDA tests for MoE and tiled attention decoding, improve FP8 and BF16 stability checks

haricot committed:
fix rebase

haricot committed:
fix rebase

haricot committed:
test(cuda): add BF16, FP8, and MoE regression coverage

haricot committed:
use cudaforge 0.1.6 (patch.crates-io) for kernel build watch conf

haricot committed:
fix rebase

dedmen commented:
Trying to use Candle on a GTX 1070, and I was struggling so much (First time rust user, first time Candle user).
CUDA Toolkit version 12.9.
Thanks to this, I got it all up and running!

All that was needed was
```
[patch.crates-io]
candle-core = { git = 'https://github.com/haricot/candle.git', branch = 'bf16_candle'}
candle-nn = { git = 'https://github.com/haricot/candle.git', branch = 'bf16_candle'}
candle-transformers = { git = 'https://github.com/haricot/candle.git', branch = 'bf16_candle'}
```
in my Cargo.toml

It would be much nicer if this PR would finally get merged :/

ivarflakstad commented:
I understand your frustration.

There are a lot of changes in this branch that we definitely want merged into main, but in that lies the issue - there are _a lot of changes_

Reviewing and verifying this is not easy. "It works on my machine" is unfortunately not good enough, because it may break what previously worked on another machine.

There are type shims both at cuda compile time and detected at runtime (do we need both?), new code paths to launch kernels, I see a moe kernel from vllm. There's even some new cuda types to aid in vectorized loading - not a bad idea to have but seems a but random to include a performance optimization in this PR?

A hundred moving parts, a hundred ways to break.

But! I think having this branch as the hub where backwards cuda compatibility is developed - and then extract only the required code into isolated PRs is a good way to get the improvements merged.

haricot commented:
Reopening this PR after rebuilding it from current main.

I tried to reduce the scope compared to the previous version. The old branch had BF16, FP8 and some other legacy CUDA changes mixed together, which made the PR harder to review.

This version only keeps the BF16 part for now.

I added an opt-in `cuda-legacy-bf16` feature. On sm_80+ the existing native BF16 path is unchanged. On older GPUs the compatibility code is only enabled when the feature is used.

I also changed the implementation quite a bit compared to the old branch. With recent CUDA 12.x, a lot of `__nv_bfloat16` conversions and basic arithmetic already work on older architectures, so instead of emulating everything I only added the few missing pieces Candle still needs, mainly BF16 `atomicAdd` and a few helper functions.

I tested this with CUDA 12.9 on a 1080 GPU.

Runtime tests currently pass for:

- casts and basic BF16 arithmetic
- min/max
- affine and unary ops
- reductions
- BF16 atomicAdd fallback
- BF16 matmul through the existing cuBLAS path with FP32 accumulation

I also checked the build with the feature disabled on sm_61, and with the feature enabled on sm_80.

I did not bring FP8 and the other changes from the previous branch back into this PR. I kept the old work in [bf16_candle_sav](https://github.com/haricot/candle/tree/bf16_candle_sav)   and, if useful, I can split the remaining parts into separate PRs later, for example:

- cuda: add opt-in FP8 storage compatibility for legacy GPUs
- cuDNN conv fallback when an architecture/operation is not supported
- pre-Volta MoE fallback 
- architecture-specific kernel/build selection
- tiled attention support, if it is still useful

Some of these parts may also need to be redesigned against current main rather than directly rebased.

Hopefully this version is easier to review than the previous one.

haricot commented:
Small update after splitting the old work into separate branches.

I made a local self-hosted runner on my Pascal `sm_61` GPU with CUDA 12.9 and cuDNN 9.10.2. Here is the successful run:

https://github.com/haricot/candle/actions/runs/32969859569/job/98180784549

The results are:

* BF16 legacy tests: 4/4 passed
* FP8 legacy tests: 4/4 passed
* CUDA convolution tests without cuDNN: 18/18 passed
* CUDA convolution tests with cuDNN: 18/18 passed

For this test I used [cuda_sm61_runner](https://github.com/haricot/candle/tree/cuda_sm61_runner), which only groups the different branches for testing. I do not plan to propose this integration branch as one PR.

After this BF16 PR, the other work is now separated like this:

* [fp8_candle](https://github.com/haricot/candle/tree/fp8_candle): legacy FP8 support
* [cudnn_fallback_candle](https://github.com/haricot/candle/tree/cudnn_fallback_candle): fallback to the existing Candle CUDA convolution when cuDNN returns `NOT_SUPPORTED` or `ARCH_MISMATCH`
* [conv2d_noncontiguous_kernel_candle](https://github.com/haricot/candle/tree/conv2d_noncontiguous_kernel_candle): a small independent fix for non-contiguous conv2d kernels
* [moe_simt_f16_candle](https://github.com/haricot/candle/tree/moe_simt_f16_candle): pre-Volta MoE fallback using SIMT F16
* [lfm25_26b_candle](https://github.com/haricot/candle/tree/lfm25_26b_candle): separate LFM model support work

I rebuilt these changes from the current Candle main instead of cherry-picking the old combined branch. I will try to propose them one by one so each PR stays small.

I also found why cuDNN was first returning `CUDNN_STATUS_EXECUTION_FAILED` on my GPU. I had cuDNN 9.11 installed, but this version no longer supports Pascal. With cuDNN 9.10.2 the same convolution tests all pass.

For this reason I kept `EXECUTION_FAILED` as a normal propagated error in the cuDNN fallback branch. Only `NOT_SUPPORTED` and `ARCH_MISMATCH` trigger the Candle CUDA fallback.

sempervictus commented:
@guoqingbao got fp4 working too, and WELL on sm70 at least
